sfge/src/main/java/com/salesforce/graph/ops/expander/ApexPathExpanderUtil.java

```diff
@@ -345,27 +326,157 @@ private void expand(
+     * as needed, and perform a true depth-first expansion.
+     */
+    private static class PendingForkStack {
+        /**
+         * We use a Stack instead of a List to allow for a true depth-first expansion. i.e., when a
+         * new fork is encountered, it's added to the top of the stack, meaning it gets processed
+         * first.
+         */
+        private final Stack<PendingForkStackFrame> stack;
+        /** The collapser is present to allow each stack frame to handle forks. */
+        private final ApexPathCollapser apexPathCollapser;
+
+        PendingForkStack(ApexPathCollapser apexPathCollapser) {
+            stack = new Stack<>();
+            this.apexPathCollapser = apexPathCollapser;
+        }
+
+        /**
+         * Use the provide Exception to create a new {@link PendingForkStackFrame} and add it to the
+         * top of the stack.
+         *
+         * @param ex An exception thrown when a method is determined to fork.
+         */
+        void addStackFrame(MethodPathForkedException ex) {
+            stack.push(new PendingForkStackFrame(ex, apexPathCollapser));
+        }
+
+        /**
+         * @return An Optional containing either the next {@link ApexPathExpander} to be processed,
+         *     or nothing.
+         */
+        Optional<ApexPathExpander> getNextExpander() {
+            // Base Case: The stack is empty, so there are no more frames to inspect.
+            // Return an empty optional.
+            if (stack.empty()) {
+                return Optional.empty();
+            }
+            PendingForkStackFrame frame = stack.peek();
+            Optional<ApexPathExpander> nextExpanderOptional = frame.getNextExpander();
+            if (nextExpanderOptional.isEmpty()) {
+                // Recursive Case: The current frame is empty. Pop it off, and call recursively
+                // to move to the next frame.
+                stack.pop();
+                return getNextExpander();
+            } else {
+                // Base Case: The current frame returned an Expander. Return it.
+                return nextExpanderOptional;
+            }
+        }
+    }
+
+    /**
+     * Each frame in the {@link PendingForkStack} represents a fork event currently being processed.
+     */
+    private static class PendingForkStackFrame {
+        /** The exception that triggered the fork. */
+        private final MethodPathForkedException ex;
+        /** The collapser used to process forks. */
+        private final ApexPathCollapser apexPathCollapser;
+        /**
+         * The index of the next {@link ApexPath} that should be turned into an {@link
+         * ApexPathExpander}.
+         */
+        private int idx;
+
+        PendingForkStackFrame(MethodPathForkedException ex, ApexPathCollapser apexPathCollapser) {
+            this.ex = ex;
+            this.apexPathCollapser = apexPathCollapser;
+            this.idx = 0;
```

Comment:
The old system generated `ApexPathExpander`s for each path and added them to the stack in order, so they were being popped off the stack and processed in reverse-order.
For the sake of readability, the new system starts at the front and processes them in order instead of reverse-order.
This has the following impacts:
- If a row previously threw an `InternalExecutionError`, it might now throw a different one or timeout instead, if it encounters a different problematic vertex or fails to reach the old one in time.
- if a row was previously timing out, it might now throw an `InternalExecutionError`, as a problematic vertex that was previously buried too deep in the callstack is now reached earlier.
If either of these are considered problems, I can just invert the loop and perfectly preserve the old behavior. I have no strong preference one way or the other.